Improve `git pull` & `git push` quick-open-items

vince-fugnitto opened the pull request:
Updated `git pull` and `git push` quick open items to now also display remote uri, instead
of simply displaying the remote name.

- create new method `getRemoteUris` to not break api, and provide more information than the standard `getRemotes` which only returns the remote name.

<div align='center'>

<img width="1049" alt="screen shot 2019-01-29 at 5 43 26 pm" src="https://user-images.githubusercontent.com/40359487/51945758-89585680-23ed-11e9-94ed-ec86df4321e2.png">

</div>


Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

vince-fugnitto commented:
Sure @kittaakos, I can make the change, my only worry is however that based on the verbosity we are returning different things. With verbosity off we are only returning remote names and with it on we are only returning the remote URI. Maybe it'll be fine, I'll just update the documentation to reflect that so there's no confusion. What do you think?

kittaakos commented:
> 'll just update the documentation to reflect that so there's no confusion

I am fine with it. Clients can parse the verbose string. Or you can return with a `Remote` object if `verbose` is `true`. Note, the `Remote` object does not exist yet.

vince-fugnitto commented:
> > 'll just update the documentation to reflect that so there's no confusion
> 
> I am fine with it. Clients can parse the verbose string. Or you can return with a `Remote` object if `verbose` is `true`. Note, the `Remote` object does not exist yet.

I like the idea of having the `Remote` object, and the clients can use the data as they choose so I'll go forward with that implementation.

vince-fugnitto commented:
@kittaakos I updated the PR but was unable to find a way to update the existing `getRemotes` method without breaking API. Since in my case I am returning the `Remote` object, it would break existing clients who are expecting an array of string instead. The only solution I could think of is to have an alternative function returning the `Remote` objects, and mark the old function for deprecation since it only returns a subset of remote information.

kittaakos commented:
Cannot you do something similar we did for `branch`?

https://github.com/theia-ide/theia/blob/1d4138c7f14ca674eaee41db8d38a3a5eae95c88/packages/git/src/common/git.ts#L603-L622

vince-fugnitto commented:
@kittaakos I updated the pr based on your comments and in a way not to break existing remote api :)

kittaakos commented on packages/git/src/common/git-model.ts:
```diff
@@ -191,6 +191,27 @@ export namespace Repository {
+/**
+ * Available Git remote actions.
+ */
+export type RemoteAction = 'fetch' | 'push';
```
The belongs to `git.ts` and not the model.

kittaakos commented on packages/git/src/common/git-model.ts:
```diff
@@ -191,6 +191,27 @@ export namespace Repository {
+/**
+ * Representation of a Git remote.
+ */
+export interface Remote {
```
```ts
interface Remote {
  name: string;
  fetch: string;
  push: string;
}
```

Thoughts?

vince-fugnitto replied:
Yes ! I think that's much nicer thank you!

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -694,6 +697,9 @@ export class DugiteGit implements Git {
 
+    /**
+     * @deprecated use `getGitRemotes` instead.
+     */
```
You can delete `private` methods.

kittaakos commented on packages/git/src/common/git.ts:
```diff
@@ -691,12 +691,20 @@ export interface Git extends Disposable {
+     * @param repository the repository to get the remote objects.
+     * @param type the git action either `fetch` or `push`.
+     */
+    remote(repository: Repository, options: { action: RemoteAction }): Promise<Remote[]>;
```
Is it possible to do something like this?
```diff
diff --git a/packages/git/src/common/git.ts b/packages/git/src/common/git.ts
index 1dc87965..a5040c34 100644
--- a/packages/git/src/common/git.ts
+++ b/packages/git/src/common/git.ts
@@ -275,6 +275,18 @@ export namespace Git {

         }

+        /**
+         * Options for `git remote`.
+         */
+        export interface Remote {
+
+            /**
+             * Be a little more verbose and show remote URL after name.
+             */
+            readonly verbose?: true;
+
+        }
+
         /**
          * Options for further refining the `git show` command.
          */
@@ -696,6 +708,9 @@ export interface Git extends Disposable {
      * @param repository the repository to get the remotes.
      */
     remote(repository: Repository): Promise<string[]>;
+    // tslint:disable-next-line:no-any
+    remote(repository: Repository, options?: Git.Options.Remote): Promise<any[]>;
+    remote(repository: Repository, options: { verbose: true }): Promise<Remote[]>;

     /**
      * Executes the Git command and resolves to the result. If an executed Git command exits with a code that is not in the `successExitCodes` or an error not in `expectedErrors`,
```

We could omit the one with the `any` return type.

kittaakos commented on packages/git/src/browser/git-quick-open-service.ts:
```diff
@@ -160,11 +164,11 @@ export class GitQuickOpenService {
-            const remotes = await this.getRemotes();
-            const defaultRemote = remotes[0]; // I wish I could use assignment destructuring here. (GH-413)
-            const executeRemote = async (remoteItem: GitQuickOpenItem<string>) => {
+            const [remotes] = await Promise.all([this.getRemotesVerbose('fetch')]);
```
Why is the `Promise.all?

vince-fugnitto replied:
Sorry was leftover from an earlier version.

kittaakos commented on packages/git/src/common/git.ts:
```diff
@@ -28,6 +28,11 @@ export const GitPath = '/services/git';
+/**
+ * Available Git remote actions.
+ */
+export type RemoteAction = 'fetch' | 'push';
```
Can be removed.

kittaakos commented on packages/git/src/common/git.ts:
```diff
@@ -546,6 +551,18 @@ export namespace Git {
+        /**
+         * Options for the `git remote` command.
+         */
+        export interface RemoteType {
```
Should be `Remote`.

kittaakos commented on packages/git/src/common/git.ts:
```diff
@@ -691,12 +708,20 @@ export interface Git extends Disposable {
+     * It resolves to an array of configured remote objects for the given Git action.
+     *
+     * @param repository the repository to get the remote objects.
+     * @param type the git action either `fetch` or `push`.
```
`@param type` is obsolete.

kittaakos commented on packages/git/src/common/git-model.ts:
```diff
@@ -191,6 +191,28 @@ export namespace Repository {
+    /**
+     * The name of the remote.
+     */
+    name: string,
```
Missing `readonly`.

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -694,18 +698,28 @@ export class DugiteGit implements Git {
+        if (results) {
+            const values: Remote[] = [];
+            for (let i = 0; i < results.length; i += 6) {
+                values.push({ name: results[i], fetch: results[i + 1], push: results[i + 4] });
```
Small remark: if you send the remote URL as is, you will transfer it through the wire as a text. Note, access tokens are secret. I am not sure if I ever want to see it in the logs or the UI. What does VS Code do?

vince-fugnitto replied:
Thanks @kittaakos I haven't thought about the case where `access tokens` are included in the url.
I wonder what the appropriate handling would be when it's the case (for logs, gui, and over the wire). I know that `vscode` does display this information but they probably won't encounter the same ' over the wire' issue since they probably won't communicate that way.

This issue leads me to believe we may currently have the same issue for other components such as transferring **Terminal** output over the wire but that is still to be investigated and confirmed.

If you don't believe the PR adds any additional value and risks too much regarding security I'm fine with scrapping it :) I noticed that `vscode` had additional info when displaying their remote items, so I thought it'd be a nice feature to include but not at the expense of risking security.

kittaakos replied:
I am fine with this PR, just wanted to make sure we are aware of this.

vince-fugnitto replied:
I assume will need to investigate and find an appropriate way to handle the `access token` security risk before being able to proceed with the PR.

kittaakos commented:
> I made updates to the pr based on your provided feedback, thank you!

👍  I will do the review.

> I'm not entirely sure what would be required to address the concern

I do not know it either.

> Do you still want to tackle the problem in this pr?

No, I never said we have to tackle it, all I wanted is to make sure we are all aware of it. See https://github.com/theia-ide/theia/pull/4190#discussion_r253042103.

kittaakos commented on packages/git/src/common/git.ts:
```diff
@@ -691,12 +703,20 @@ export interface Git extends Disposable {
+     * @param repository the repository to get the remote objects.
+     * @param options `git remote` command refinements.
+     */
+    remote(repository: Repository, options?: Git.Options.Remote): Promise<RemoteModel[]>;
```
`options` cannot be optional here. Otherwise, at call-site we cannot infer the return type.

vince-fugnitto replied:
fixed!

kittaakos commented on packages/git/src/common/git.ts:
```diff
@@ -691,12 +703,20 @@ export interface Git extends Disposable {
+     * @param repository the repository to get the remote objects.
+     * @param options `git remote` command refinements.
+     */
+    remote(repository: Repository, options: Git.Options.Remote): Promise<RemoteModel[]>;
```
It should be:
```ts
remote(repository: Repository, options: { verbose: true }): Promise<RemoteModel[]>;
```

With your approach I can do the following on the call-site:
```ts
await this.git.remote({ localUri: 'file:///blabla' }, {}); // {} is valid becasue `verbose` is optional. We must require an explicit `verbose = true`. Otherwise, it won't work.
```

But please correct me if I have overlooked something.

vince-fugnitto replied:
yes I agree, I've fixed it

kittaakos commented on packages/git/src/browser/git-quick-open-service.ts:
```diff
@@ -366,6 +374,16 @@ export class GitQuickOpenService {
     }
 
+    private async getRemotesVerbose(): Promise<Remote[]> {
+        const repository = this.getRepository();
```
[minor]: I would use one `getRemote` method in this class instead of `getRemotes` and `getRemotesVerbose`. It is completely OK to delete `private` fields and methods.

vince-fugnitto replied:
I've removed the unnecessary `getRemotesVerbose` call, and refactored the existing `getRemotes`.

AlexTugarev commented on packages/git/src/browser/git-quick-open-service.ts:
```diff
@@ -175,7 +179,7 @@ export class GitQuickOpenService {
                     const executeBranch = async (branchItem: GitQuickOpenItem<Branch>) => {
                         try {
-                            await this.git.pull(repository, { remote: remoteItem.ref, branch: branchItem.ref.nameWithoutRemote });
+                            await this.git.pull(repository, { remote: remoteItem.ref.name, branch: branchItem.ref.nameWithoutRemote });
```
Not sure how this happens, but this breaks the `pull` operation.

Please set a breakpoint [here](https://github.com/theia-ide/theia/blob/7ca74cd789e629f9477b8755d201b1312b17e387/packages/git/src/node/dugite-git.ts#L517) and look at the value of `currentRemote`. It supposed to be a string, but I see:
```
 {
  "name": "origin",
  "fetch": "https://github.com/theia-ide/theia.git",
  "push": "https://github.com/theia-ide/theia.git"
 }
```

vince-fugnitto replied:
@AlexTugarev 
I wasn't able to reproduce the issue :disappointed: 
I've tried both with a repository with a single remote, as well as multiple remotes but I still can't reproduce the issue about having the `Remote` object at this point instead of the string.

kittaakos replied:
@vince-fugnitto, can you try it in gitpod?

I did the followings:
 - Open the PR in Gitpod: https://gitpod.io/#https://github.com/theia-ide/theia/pull/4190
 - Wait for the build, then execute `yarn --cwd ./examples/browser/ start --hostname 0.0.0.0 --port 3000` in the terminal.
 - Click on `Open Browser`.
 - Open `theia` in the workspace.
 - Go to the Git view and `pull`.

If you have any questions, let me know. Thank you!

AlexTugarev replied:
Thanks @kittaakos!

vince-fugnitto replied:
thanks @AlexTugarev @kittaakos I'll take a look and see what the issue is

vince-fugnitto replied:
I tried again and seem to get the correct value for `currentRemote`, in this case "origin".

![image](https://user-images.githubusercontent.com/40359487/52425576-4cb1ec80-2aca-11e9-98b9-e04338cd586d.png)

Is there something I may be doing incorrectly?

kittaakos replied:
> Is there something I may be doing incorrectly?

No idea.

I had a `Request pull failed with message: Authentication failed. You may not have permission to access the repository or the repository may have been archived.` error. I have to look into it again... Bear with me!

vince-fugnitto replied:
@kittaakos did you get the chance to look into the issue again?

kittaakos replied:
I am checking it once more...

lmcbout commented:
Using "git pull from", I see what is registered in my .git/config for the remote. It could be "git@..." or "ssh://..."

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -550,10 +550,14 @@ export class DugiteGit implements Git {
-        return this.getRemotes(repositoryPath);
+        const remotes = await this.getRemotes(repositoryPath);
+        const names = remotes.map(a => a.name);
+        return (options) ? remotes : names;
```
This seems to be incorrect. We need the full `Remote` objects when `options` is defined AND `options.verbose === true`.

kittaakos commented:
I can see this when I am doing `Fetch...` from the UI:

![screen shot 2019-02-12 at 08 45 06](https://user-images.githubusercontent.com/1405703/52619682-a5a2cb80-2ea2-11e9-84e1-bcd3d7b9803e.jpg)

vince-fugnitto commented:
@kittaakos @AlexTugarev 
I've updated the PR based on your comments and findings. I believe I've fixed the issue regarding the bug you previously mentioned.

vince-fugnitto commented:
@kittaakos thank you for your help and review !
